Re-selecting an option in a drop down item with valueString answer options is causing app to crash

**Describe the bug**
An item with the type 'choice', using the drop down extension, and answer options comprised of valueStrings crashes the app when an option is re-selected. The first time the option is selected, nothing happens but when you attempt to select an other option from the same drop down, the app crashes.

**Component**
Demo app

**To Reproduce**
Steps to reproduce the behavior:
1. Update the file 'new-patient-registration-paginated.json' to use an item of the type 'choice' for the country attribute, remove the initial expression attribute (because of [this](https://github.com/google/android-fhir/issues/1541) issue), and add valueString answer options.
2. Run the demo app
3. Tap on a patient
4. Tap on the edit button
5. Select a country
6. Select a different country
7. The app will crash with the error: Caused by: java.lang.IllegalArgumentException: Mismatching question type CHOICE and answer type string for PR-address-country

**Expected behavior**
I expect the second country to be selected

**Screenshots**
N/A

**Smartphone (please complete the following information):**
- Device: Samsung A12
- Android version: 11
- Build number: RP1A.200720.012.A127FXXU4AUK1
- Version: 30

**Additional context**
Support for the string type for a choice widget was added in this [PR](https://github.com/google/android-fhir/pull/1399). The updated item would look like this:
```
{
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "coding": [
                      {
                        "system": "http://hl7.org/fhir/questionnaire-item-control",
                        "code": "drop-down",
                        "display": "Drop down"
                      }
                    ],
                    "text": "Drop down"
                  }
                }
              ],
              "linkId": "PR-address-country",
              "definition": "http://hl7.org/fhir/StructureDefinition/Patient#Patient.address.country",
              "type": "choice",
              "answerOption": [
                {
                  "valueString": "Kenya"
                },
                {
                  "valueString": "Malawi"
                },
                {
                  "valueString": "Zambia"
                }
              ],
              "required": true,
              "text": "Country",
              "_text": {
                "extension": [
                  {
                    "extension": [
                      {
                        "url": "lang",
                        "valueCode": "sw"
                      },
                      {
                        "url": "content",
                        "valueString": "Nchi"
                      }
                    ],
                    "url": "http://hl7.org/fhir/StructureDefinition/translation"
                  }
                ]
              }
            }
          ]
        }
```

**Would you like to work on the issue?**
Maybe the author of the [PR](https://github.com/google/android-fhir/pull/1399) can look into resolving this

--- a/datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactoryEspressoTest.kt
+++ b/datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactoryEspressoTest.kt
@@ -36,6 +36,7 @@ import com.google.common.truth.Truth.assertThat
 import org.hl7.fhir.r4.model.Coding
 import org.hl7.fhir.r4.model.Questionnaire
 import org.hl7.fhir.r4.model.QuestionnaireResponse
+import org.hl7.fhir.r4.model.StringType
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
@@ -96,6 +97,28 @@ class QuestionnaireItemDropDownViewHolderFactoryEspressoTest {
       .isEqualTo("Coding 3")
   }
 
+  @Test
+  fun shouldSetDropDownValueStringToAutoCompleteTextView() {
+    val questionnaireItemViewItem =
+      QuestionnaireItemViewItem(
+        answerOptionsValueString("Coding 1", "Coding 2", "Coding 3", "Coding 4", "Coding 5"),
+        responseValueStringOptions(),
+        validationResult = NotValidated,
+        answersChangedCallback = { _, _, _ -> },
+      )
+    runOnUI { viewHolder.bind(questionnaireItemViewItem) }
+
+    onView(withId(R.id.auto_complete)).perform(showDropDown())
+    onView(withText("Coding 1"))
+      .inRoot(isPlatformPopup())
+      .check(matches(isDisplayed()))
+      .perform(click())
+    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.auto_complete).text.toString())
+      .isEqualTo("Coding 1")
+    assertThat((questionnaireItemViewItem.answers.single().value as StringType).valueAsString)
+      .isEqualTo("Coding 1")
+  }
+
   /** Method to run code snippet on UI/main thread */
   private fun runOnUI(action: () -> Unit) {
     activityScenarioRule.scenario.onActivity { action() }
@@ -128,6 +151,28 @@ class QuestionnaireItemDropDownViewHolderFactoryEspressoTest {
         addAnswer(
           QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
             value = Coding().apply { display = response }
+          }
+        )
+      }
+    }
+
+  private fun answerOptionsValueString(vararg options: String) =
+    Questionnaire.QuestionnaireItemComponent().apply {
+      options.forEach { option ->
+        addAnswerOption(
+          Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
+            value = StringType().apply { valueAsString = option }
+          }
+        )
+      }
+    }
+
+  private fun responseValueStringOptions(vararg responses: String) =
+    QuestionnaireResponse.QuestionnaireResponseItemComponent().apply {
+      responses.forEach { response ->
+        addAnswer(
+          QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+            value = StringType().apply { valueAsString = response }
           }
         )
       }

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt
@@ -22,6 +22,7 @@ import org.hl7.fhir.r4.model.Coding
 import org.hl7.fhir.r4.model.DateType
 import org.hl7.fhir.r4.model.IntegerType
 import org.hl7.fhir.r4.model.Questionnaire
+import org.hl7.fhir.r4.model.QuestionnaireResponse
 import org.hl7.fhir.r4.model.StringType
 import org.hl7.fhir.r4.model.TimeType
 
@@ -33,6 +34,27 @@ internal const val EXTENSION_OPTION_EXCLUSIVE_URL =
  * option is [IntegerType] or [StringType] or [Coding] type.
  */
 internal val Questionnaire.QuestionnaireItemAnswerOptionComponent.displayString: String
+  get() {
+    return when (value) {
+      is IntegerType, is DateType, is TimeType -> value.primitiveValue()
+      is StringType -> (value as StringType).getLocalizedText() ?: value.toString()
+      is Coding -> {
+        val display = valueCoding.displayElement.getLocalizedText() ?: valueCoding.display
+        if (display.isNullOrEmpty()) {
+          valueCoding.code
+        } else {
+          display
+        }
+      }
+      else -> throw IllegalArgumentException("$value is not supported.")
+    }
+  }
+
+/**
+ * Text value for response item answer option [QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent] if answer
+ * option is [IntegerType] or [StringType] or [Coding] type.
+ */
+internal val QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.displayString: String
   get() {
     return when (value) {
       is IntegerType, is DateType, is TimeType -> value.primitiveValue()

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/MoreAnswerOptions.kt
@@ -51,8 +51,9 @@ internal val Questionnaire.QuestionnaireItemAnswerOptionComponent.displayString:
   }
 
 /**
- * Text value for response item answer option [QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent] if answer
- * option is [IntegerType] or [StringType] or [Coding] type.
+ * Text value for response item answer option
+ * [QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent] if answer option is
+ * [IntegerType] or [StringType] or [Coding] type.
  */
 internal val QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.displayString: String
   get() {

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt
@@ -276,7 +276,7 @@ object QuestionnaireResponseValidator {
           "Mismatching question type $questionnaireItemType and answer type $answerType for $linkId"
         }
       Questionnaire.QuestionnaireItemType.CHOICE ->
-        require(answerType == "Coding") {
+        require(answerType == "Coding" || answerType == "string") {
           "Mismatching question type $questionnaireItemType and answer type $answerType for $linkId"
         }
       Questionnaire.QuestionnaireItemType.OPENCHOICE ->

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt
@@ -275,11 +275,7 @@ object QuestionnaireResponseValidator {
         require(answerType == "url") {
           "Mismatching question type $questionnaireItemType and answer type $answerType for $linkId"
         }
-      Questionnaire.QuestionnaireItemType.CHOICE ->
-        require(answerType == "Coding" || answerType == "string") {
-          "Mismatching question type $questionnaireItemType and answer type $answerType for $linkId"
-        }
-      Questionnaire.QuestionnaireItemType.OPENCHOICE ->
+      Questionnaire.QuestionnaireItemType.CHOICE, Questionnaire.QuestionnaireItemType.OPENCHOICE ->
         require(answerType == "Coding" || answerType == "string") {
           "Mismatching question type $questionnaireItemType and answer type $answerType for $linkId"
         }

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactory.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactory.kt
@@ -58,7 +58,7 @@ internal object QuestionnaireItemDropDownViewHolderFactory :
         val adapter =
           ArrayAdapter(context, R.layout.questionnaire_item_drop_down_list, answerOptionString)
         autoCompleteTextView.setText(
-          questionnaireItemViewItem.answers.singleOrNull()?.value?.toString() ?: ""
+          questionnaireItemViewItem.answers.singleOrNull()?.displayString ?: ""
         )
         autoCompleteTextView.setAdapter(adapter)
         autoCompleteTextView.onItemClickListener =

--- a/datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactory.kt
+++ b/datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDropDownViewHolderFactory.kt
@@ -57,7 +57,7 @@ internal object QuestionnaireItemDropDownViewHolderFactory :
         val adapter =
           ArrayAdapter(context, R.layout.questionnaire_item_drop_down_list, answerOptionString)
         autoCompleteTextView.setText(
-          questionnaireItemViewItem.answers.singleOrNull()?.valueCoding?.display ?: ""
+          questionnaireItemViewItem.answers.singleOrNull()?.value?.toString() ?: ""
         )
         autoCompleteTextView.setAdapter(adapter)
         autoCompleteTextView.onItemClickListener =

